File: src/main/java/com/atsuishio/superbwarfare/item/BatteryItem.java
package com.atsuishio.superbwarfare.item;

import com.atsuishio.superbwarfare.client.tooltip.component.CellImageComponent;
import net.minecraft.world.inventory.tooltip.TooltipComponent;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.energy.EnergyStorage;
import net.neoforged.neoforge.energy.IEnergyStorage;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

public class BatteryItem extends Item {

    public int maxEnergy;

    public IEnergyStorage getEnergyStorage() {
        return energy;
    }

    private final IEnergyStorage energy;

    public BatteryItem(int maxEnergy, Properties properties) {
        super(properties.stacksTo(1));
        this.maxEnergy = maxEnergy;
        this.energy = new EnergyStorage(maxEnergy);
    }

    @Override
    public boolean isBarVisible(ItemStack pStack) {
        var cap = pStack.getCapability(Capabilities.EnergyStorage.ITEM);
        if (cap == null) return false;
        return cap.getEnergyStored() != cap.getMaxEnergyStored();
    }

    @Override
    public int getBarWidth(ItemStack pStack) {
        var energy = 0;
        var cap = pStack.getCapability(Capabilities.EnergyStorage.ITEM);
        if (cap != null) {
            energy = cap.getEnergyStored();
        }

        return Math.round((float) energy * 13.0F / maxEnergy);
    }

    @Override
    public int getBarColor(@NotNull ItemStack pStack) {
        return 0xFFFF00;
    }

    @Override
    public @NotNull Optional<TooltipComponent> getTooltipImage(@NotNull ItemStack pStack) {
        return Optional.of(new CellImageComponent(pStack));
    }

    public ItemStack makeFullEnergyStack() {
        ItemStack stack = new ItemStack(this);
        var cap = stack.getCapability(Capabilities.EnergyStorage.ITEM);
        if (cap == null) return stack;

        cap.receiveEnergy(maxEnergy, false);
        return stack;
    }

}

File: src/main/kotlin/com/atsuishio/superbwarfare/item/material/BatteryItem.kt
package com.atsuishio.superbwarfare.item.material

import com.atsuishio.superbwarfare.client.tooltip.component.CellImageComponent
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.tooltip.TooltipComponent
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.neoforged.neoforge.capabilities.Capabilities
import java.util.*
import kotlin.math.roundToInt

open class BatteryItem(var maxEnergy: Int, properties: Properties) : Item(properties.stacksTo(1)) {
    override fun isBarVisible(pStack: ItemStack): Boolean {
        val cap = pStack.getCapability(Capabilities.EnergyStorage.ITEM) ?: return false
        return cap.energyStored != cap.maxEnergyStored
    }

    override fun getBarWidth(pStack: ItemStack): Int {
        var energy = 0
        val cap = pStack.getCapability(Capabilities.EnergyStorage.ITEM)
        if (cap != null) {
            energy = cap.energyStored
        }

        return (energy * 13f / maxEnergy).roundToInt()
    }

    override fun getBarColor(pStack: ItemStack) = 0xFFFF00

    override fun getTooltipImage(pStack: ItemStack): Optional<TooltipComponent> {
        return Optional.of(CellImageComponent(pStack))
    }

    fun makeFullEnergyStack(): ItemStack {
        val stack = ItemStack(this)
        val cap = stack.getCapability(Capabilities.EnergyStorage.ITEM) ?: return stack

        cap.receiveEnergy(maxEnergy, false)
        return stack
    }

    override fun inventoryTick(pStack: ItemStack, pLevel: Level, entity: Entity, pSlotId: Int, pIsSelected: Boolean) {
        super.inventoryTick(pStack, pLevel, entity, pSlotId, pIsSelected)
        // TODO 完成寻找能充电的物品进行充电（包括饰品栏里的？）
        if (entity is Player) {
            for (chargeable in entity.inventory.items) {
                if (chargeable.item is BatteryItem) {
//                    val stackStorage = chargeable.getCapability(Capabilities.EnergyStorage.ITEM);
//
//                    assert(stack.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).resolve().isPresent())
//                    val stackStorage: IEnergyStorage =
//                        stack.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).resolve().get()
//                    val stackMaxEnergy = stackStorage.maxEnergyStored
//                    val stackEnergy = stackStorage.energyStored
//
//                    assert(cell.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).resolve().isPresent)
//                    val cellStorage = cell.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).resolve().get()
//                    val cellEnergy = cellStorage.energyStored
//
//                    val stackEnergyNeed =
//                        min(cellEnergy.toDouble(), (stackMaxEnergy - stackEnergy).toDouble()).toInt()
//
//                    if (cellEnergy > 0) {
//                        stack.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).ifPresent(
//                            NonNullConsumer<IEnergyStorage> { iEnergyStorage: IEnergyStorage ->
//                                iEnergyStorage.receiveEnergy(
//                                    stackEnergyNeed,
//                                    false
//                                )
//                            }
//                        )
//                    }
//                    cell.getCapability<IEnergyStorage>(ForgeCapabilities.ENERGY).ifPresent { cEnergy: IEnergyStorage ->
//                        cEnergy.extractEnergy(
//                            stackEnergyNeed,
//                            false
//                        )
//                    }
                }
            }
        }
    }
}